[OMPIRBuilder] Don't discard the debug record from entry block.

abidh opened the pull request:
When we get a function back from `CodeExtractor`, we discard its entry block after coping its instructions into the entry block we prepared. While copying the instructions, the terminator is discarded for obvious reasons. But if there were some debug values attached to the terminator, those are useful and needs to be copied.

abidh committed:
[OMPIRBuilder] Don't discard the debug record from entry block.

When we get a function back from CodeExtractor, we disard its entry
block after coping its instructions into the entry block we prepared.
While copying the instructions, the terminator is discarded for obvious
reasons. But if there were some debug values attached to the terminator,
those are useful and needs to be copied.

Meinersbur reviewed (commented):
Could you add a test so we do not regress in the future?

abidh commented:
> Could you add a test so we do not regress in the future?

Hi @Meinersbur 

Thanks for taking a look. Although this change fixes a missing piece, there is currently no way to test it as `CodeExtractor` does not generate the debug info in the entry block. This is preparatory PR for the bigger changes I am working on to improve the debug info of the function generated by the CodeExtractor. The bigger PR will have the tests.

This piece was independent and self contained so I thought I will post it separately.

abidh committed:
Update clang test that use OMPIRBuilder for parallel regions.

abidh committed:
Add mlir->llvmir test.

abidh commented:
We can test this functionality now as  https://github.com/llvm/llvm-project/pull/136016 has been merged. I have added a test and updated relevant clang tests.

Meinersbur reviewed (approved):
LGTM, thank you.
